rfcs/45-stageless.md

```diff
@@ -0,0 +1,537 @@
+
+## Implementation strategy
+
+Let's take a look at what implementing this would take:
+
+1. Completely rip out:
+   1. `Stage`
+   2. `Schedule`
+   3. `SystemDescriptor`
+   4. `IntoExclusiveSystem`
+2. Build out a basic schedule abstraction:
+   1. Create a `ScheduleLabel` trait
+   2. Store multiple schedules in an `App` using a dictionary approach
+   3. Support adding systems to the schedule
+      1. These should be initialized by definition: see [PR #2777](https://github.com/bevyengine/bevy/pull/2777)
+   4. Support adding systems to specific schedules and create startup system API
+   5. Special-cased data access control for "the entire world" and `Schedules` to support exclusive systems
+3. Rebuild core command processing machinery
+   1. Begin with trivial event-style commands
+   2. Explore how to regain parallelism
+4. Add system configuration
+   1. Create `SystemConfig` type
+   2. Systems labels store a `SystemConfig`
+   3. Allow systems to be labelled
+   4. Store `ConfiguredSystems` in the schedule
+   5. Add `.add_system_set` method
+   6. Use [system builder syntax](https://github.com/bevyengine/rfcs/pull/31), rather than adding more complex methods to `App`
+5. Add basic ordering constraints: basic data structures and configuration methods
+   1. Begin with strict ordering constraints: simplest and most fundamental
+   2. Add if-needed ordering constraints
+6. Add atomic ordering constraints
+7. Add run criteria
+   1. Create `IntoRunCriteriaSystem` machinery
+   2. Store the run criteria of each `ConfiguredSystem` in the schedule
+   3. Add atomic ordering constraints
+   4. Check the values of the run-criteria of systems before deciding whether it can be run
+   5. Add methods to configure run criteria
+8. Add states
+   1. Create `State` trait
+   2. Implement on-update states as sugar for simple run criteria
+   3. Create on-enter and on-exit schedules
+   4. Create sugar for adding systems to these schedules
+9. Add new examples
+   1. Complex control flow with supplementary schedules
+   2. Fixed time-step pattern
+10. Port the entire engine to the new paradigm
+    1. We almost certainly need to port the improved ambiguity checker over to make this reliable
+
+Given the massive scope, that sounds relatively straightforward!
+However, doing so will break approximately the entire engine, and tests will not pass again until step 10.
+
+Lets explore some of the more critical and challenging details.
+
+### Storing and configuring systems
+
+Schedules store systems in a configured, initialized form.
+
+```rust
+struct Schedule{
+   // Each schedule is associated with exactly one `World`, as systems must be initiliazed on it
+   world_id: WorldId
+   // We need to be able to quickly look up specific systems and iterate over them
+   // `SystemId` should be a unique identifier, atomically generated on system insertion
+   systems: HashMap<SystemId, ConfiguredSystem>,

```

Comment:
Could the system id be just an integer index, based on an incrementing counter, and allow this to be a Vec rather than a HashMap?

Reply:
Hmm, you're probably right. That would be faster, but it would make the type signatures of everything else less clear.

Reply:
Actually no, we can't do this anymore: it will break everything if we ever remove systems.

Added a comment to that effect.